server.go

```diff
@@ -1521,62 +1534,115 @@ func (s *Server) ExecuteQuery(q *influxql.Query, database string, user *User) (i
 		case *influxql.RevokeStatement:
 			continue
 		case *influxql.CreateRetentionPolicyStatement:
-			return s.executeCreateRetentionPolicyStatement(c, user)
+			res = s.executeCreateRetentionPolicyStatement(stmt, user)
 		case *influxql.AlterRetentionPolicyStatement:
-			return s.executeAlterRetentionPolicyStatement(c, user)
+			res = s.executeAlterRetentionPolicyStatement(stmt, user)
 		case *influxql.DropRetentionPolicyStatement:
-			return s.executeDropRetentionPolicyStatement(c, user)
+			res = s.executeDropRetentionPolicyStatement(stmt, user)
 		case *influxql.ListRetentionPoliciesStatement:
-			return s.executeListRetentionPoliciesStatement(c, user)
+			res = s.executeListRetentionPoliciesStatement(stmt, user)
 		case *influxql.CreateContinuousQueryStatement:
 			continue
 		case *influxql.DropContinuousQueryStatement:
 			continue
 		case *influxql.ListContinuousQueriesStatement:
 			continue
+		default:
+			panic(fmt.Sprintf("unsupported statement type: %T", stmt))
 		}
+
+		// If an error occurs then stop processing remaining statements.
+		results[i] = res
+		if res.Err != nil {
+			break
+		}
+	}
+
+	// Fill any empty results after error.
+	for i, res := range results {
+		if res == nil {
+			results[i] = &Result{Err: ErrNotExecuted}
+		}
+	}
+
+	return results
+}
+
+// executeSelectStatement plans and executes a select statement against a database.
+func (s *Server) executeSelectStatement(stmt *influxql.SelectStatement, database string, user *User) *Result {
+	// Plan statement execution.
+	e, err := s.planSelectStatement(stmt, database)
+	if err != nil {
+		return &Result{Err: err}
+	}
+
+	// Execute plan.
+	ch, err := e.Execute()
+	if err != nil {
+		return &Result{Err: err}
+	}
+
+	// Read all rows from channel.
+	res := &Result{Rows: make([]*influxql.Row, 0)}
+	for row := range ch {
+		res.Rows = append(res.Rows, row)
+	}
+
+	return res
+}
+
+// plans a selection statement under lock.
+func (s *Server) planSelectStatement(stmt *influxql.SelectStatement, database string) (*influxql.Executor, error) {
+	s.mu.Lock()
```

Comment:
The lock just exists so we can safely grab the metadata. The shards won't be able to get dropped while there is a query in progress but that's because the Bolt `Tx` grabs a read lock.